### PR TITLE
Key assignment related panic cleanup

### PR DESCRIPTION
**Pull request by smarshall-spitzbart**
# Description

WIP, optionally we can audit the key related panics in `x/ccv/utils/utils.go` in this PR as well.

**_EDIT:_** 

- [x] `utils.TMCryptoPublicKeyToConsAddr` should return an error instead. 

## Linked issues

Works toward https://github.com/cosmos/interchain-security/issues/621

## Type of change

If you've checked more than one of the first three boxes, consider splitting this PR into multiple PRs!

- [ ] `Feature`: Changes and/or adds code behavior, irrelevant to bug fixes
- [ ] `Fix`: Changes and/or adds code behavior, specifically to fix a bug
- [ ] `Refactor`: Changes existing code style, naming, structure, etc.
- [ ] `Testing`: Adds testing
- [ ] `Docs`: Adds documentation

## Regression tests

If `Refactor`, describe the new or existing tests that verify no behavior was changed or added where refactors were introduced.

## New behavior tests

If `Feature` or `Fix`, describe the new or existing tests that verify the new behavior is correct and expected.

### CODE REVIEW

File: x/ccv/provider/types/genesis.go [hunk 1]
```diff
@@ -85,6 +85,13 @@ func (gs GenesisState) Validate() error {
 		return err
 	}
 
+	for _, pk := range gs.ValidatorConsumerPubkeys {
+		if err := sdk.VerifyAddressFormat(pk.ProviderAddr); err != nil {
+			return sdkerrors.Wrap(ccv.ErrInvalidGenesis, fmt.Sprintf("invalid provider address: %s", pk.ProviderAddr))
+		}
+		// TODO: do we need to validate the consumer key?
```
[the remark this answers is not in the document]
Comment (mpoke) [reply x/ccv/provider/types/genesis.go:92]: Now that I look at it, the `MsgAssignConsumerKey.ConsumerKey` should be validated as `MsgCreateValidator.Pubkey`, while `MsgAssignConsumerKey.ProviderAddr` as `MsgCreateValidator.ValidatorAddress`, see https://github.com/cosmos/cosmos-sdk/blob/47f46643affd7ec7978329c42bac47275ac7e1cc/x/staking/types/msg.go#L90.

File: x/ccv/provider/keeper/key_assignment.go [hunk 2]
```diff
@@ -120,10 +133,8 @@ func (k Keeper) SetValidatorByConsumerAddr(
 	providerAddr sdk.ConsAddress,
 ) {
 	store := ctx.KVStore(k.storeKey)
-	bz, err := providerAddr.Marshal()
-	if err != nil {
-		panic(err)
-	}
+	// Cons address is a type alias for a byte string, no marshaling needed
+	bz := providerAddr
```
Comment (mpoke) [inline x/ccv/provider/keeper/key_assignment.go:137]: Nice

File: x/ccv/provider/keeper/key_assignment.go [hunk 3]
```diff
@@ -272,12 +297,17 @@ func (k Keeper) AppendConsumerAddrsToPrune(ctx sdk.Context, chainID string, vscI
 	if bz != nil {
 		err := consumerAddrsToPrune.Unmarshal(bz)
 		if err != nil {
+			// An error here would indicate something is very wrong,
+			// the data bytes are assumed to be correctly serialized by previous calls to this method.
 			panic(err)
 		}
 	}
+	// TODO: do we need to validate the consumerAddr we're appending here?
```
Comment (mpoke) [inline x/ccv/provider/keeper/key_assignment.go:305]: `AppendConsumerAddrsToPrune` is called either in `AssignConsumerKey` on `DeliverTx` or in `InitGenesis`.

In `AssignConsumerKey`, `consumerAddr` will be correct as long as we validate all assigned `consumerKeys`. Note that the `utils.TMCryptoPublicKeyToConsAddr`. Maybe we should change this function to return an error. 
```golang
oldConsumerKey := k.GetValidatorConsumerPubKey(ctx, chainID, providerAddr)
consumerAddr := utils.TMCryptoPublicKeyToConsAddr(oldConsumerKey)
```

In `InitGenesis`, `consumerAddr` comes `genState.ConsumerAddrsToPrune` which should be validated by `GenesisState.Validate()`, but it's not.

File: x/ccv/provider/keeper/key_assignment.go [hunk 4]
```diff
@@ -40,7 +44,10 @@ func (k Keeper) SetValidatorConsumerPubKey(
 	store := ctx.KVStore(k.storeKey)
 	bz, err := consumerKey.Marshal()
 	if err != nil {
-		panic(err)
+		// An error here would indicate something is very wrong,
+		// the consumer key is obtained from GetValidatorConsumerPubKey, called from
+		// TODO: not sure that consumerKey is actually validated by all possible code paths
```
Comment (mpoke) [inline x/ccv/provider/keeper/key_assignment.go:49]: Me neither. @danwt could you please verify that.

File: x/ccv/provider/types/genesis.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (danwt) [reply x/ccv/provider/types/genesis.go:92]: Re MsgAssignConsumerKey.ValidateBasic I was following the pattern in the very similar `MsgCreateValidator` from sdk staking. The only thing about pubkey that they check is that it's not nil.

https://github.com/cosmos/cosmos-sdk/blob/f771f20da49afc0fa3e0e75da431733382ddf3dd/x/staking/types/msg.go#L103-L105

Re MsgCreateValidator.Pubkey, I failed to match their pattern

https://github.com/cosmos/cosmos-sdk/blob/f771f20da49afc0fa3e0e75da431733382ddf3dd/x/staking/types/msg.go#L95-L98

Good spot. I've amended in

https://github.com/cosmos/interchain-security/pull/644

**Comment by danwt** [comment]
Nice job @smarshall-spitzbart!
I've got to check some stuff myself to resolve the remaining TODOs

File: x/ccv/provider/types/genesis.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (smarshall-spitzbart) [reply x/ccv/provider/types/genesis.go]: @danwt I've removed my genesis validation code in favor of https://github.com/cosmos/interchain-security/pull/646, the complete solution. That PR will resolve this comment, correct?

File: x/ccv/provider/keeper/key_assignment.go (hunk 3 above)
Comment (smarshall-spitzbart) [reply x/ccv/provider/keeper/key_assignment.go]: It seems like the pruned consumerAddr is indeed validated by all calling functions then? See https://github.com/cosmos/interchain-security/commit/ba19b5818e249fdc9f4c31a500a9625bd928859f which removes the TODO comment.

If you think it's worth changing `utils.TMCryptoPublicKeyToConsAddr` to return an error, I can do so 👍. Note [this line](https://github.com/cosmos/interchain-security/blob/ba19b5818e249fdc9f4c31a500a9625bd928859f/x/ccv/provider/keeper/grpc_query.go#L106) could panic during query for example. Hopefully the version of sdk we use allows the binaries to recover from panics during queries, but I'm not sure

I believe the last part of your comment re genesis validation should be handled by https://github.com/cosmos/interchain-security/pull/646.

File: x/ccv/provider/keeper/key_assignment.go (hunk 4 above)
Comment (smarshall-spitzbart) [reply x/ccv/provider/keeper/key_assignment.go:49]: For InitGenesis, it seems consumerKey is validated properly once #646 is merged. The only other calling function is `AssignConsumerKey`. 

If the consumer key from the [msg server method](https://github.com/cosmos/interchain-security/blob/ba19b5818e249fdc9f4c31a500a9625bd928859f/x/ccv/provider/keeper/msg_server.go#L29) is validated properly, I believe this TODO can be resolved

**Comment by smarshall-spitzbart** [comment]
@mpoke @danwt I believe with #646 merged this PR should just be comments? The PR seems ready to review tho

File: x/ccv/utils/utils.go [hunk 5]
```diff
@@ -53,12 +53,14 @@ func GetChangePubKeyAddress(change abci.ValidatorUpdate) (addr []byte) {
 	return pk.Address()
 }
 
-func TMCryptoPublicKeyToConsAddr(k tmprotocrypto.PublicKey) sdk.ConsAddress {
+// TMCryptoPublicKeyToConsAddr converts a TM public key to an SDK public key
+// and returns the associated consensus address
+func TMCryptoPublicKeyToConsAddr(k tmprotocrypto.PublicKey) (sdk.ConsAddress, error) {
```
Comment (mpoke) [inline x/ccv/utils/utils.go:58]: @smarshall-spitzbart I changed this function to return error instead of panicking.

**Review by mpoke — APPROVED** [review]

**Review by danwt — APPROVED** [review]

**Comment by smarshall-spitzbart** [comment]
Re-reviewed myself and LGTM

**Pull request merged**